### ISSUE TITLE
library routine called out of sequence

### ISSUE THREAD

**Issue by FrankyShy** [Tencent/wcdb#420]
### The language of WCDB

> e.g. Objective-C, Swift or Java 

Objective-C

### The version of WCDB

> e.g. v1.0.5

v1.0.7

### The platform of WCDB

> e.g. iOS, macOS or Android

iOS

### The installation of WCDB

> e.g. Cocoapods, Carthage, Maven, AAR Package or Git clone

Cocoapods

### What's the issue?

> Post the outputs or screenshots for errors.
> 
> Explain what you want by example or code **in English**.

<img width="1100" alt="fc50a5a3-88df-40d6-9a04-832ffe41307f" src="https://user-images.githubusercontent.com/15244415/48329392-9170e780-e683-11e8-8c47-8c5afd97c2f7.png">

when we use the WCDB.framework & sqlcipher.framework, we find the fmdb library method report a error with "library routine called out of sequence.".but  when we use the source pod,the project run normally.

[this is demo](https://pan.baidu.com/s/1E8m536fys7eBIKsYCtDTsw)

"WCDBDemo-Framework.zip" is demo with  WCDB.framework & sqlcipher.framework.
"WCDBDemo-SCCode.zip" is demo with the source pod.

**Comment by lsqswl** [comment]
I have the same question. Our project cannot allow me to add "use_framworks!" becaurse it will make other third library cause more questions.

**Comment by lsqswl** [comment]
https://github.com/Tencent/wcdb/issues/342

**Comment by FrankyShy** [comment]
I try to add "use_frameworkes!" to podfile, the issue is also exist. My project don't add the other SQLite frameworks. Please help me to see the demo in baidu clouddisk.
@RingoD

**Comment by FrankyShy** [comment]
The WCDB's function is normal,but the fmdb occur errors.

**Comment by RingoD** [comment]
@lsqswl  Then you can use the solution 2.

**Comment by RingoD** [comment]
@FrankyShy  Did you run `pod deintegrate` and `pod install` to re-install the frameworks? In additional, you can push you project.zip here directly, since we don't use baidu cloud.

**Comment by RingoD** [comment]
@FrankyShy 
Why your podfile is so strange?
```
pod 'WCDB', '1.0.7-beta9-TEST'
pod 'WCDBOptimizedSQLCipher', '0.0.1-beta3-TEST'
```
Why did you use the standard WCDB pod with v1.0.7.5?

**Comment by FrankyShy** [comment]
"WCDB 1.0.7-beta9-TEST" and "WCDBOptimizedSQLCipher  0.0.1-beta3-TEST" are frameworks from your project,we just put at our company Mirrors.
We don't use WCDB v1.0.7.5,because it's source code not framework.

**Comment by RingoD** [comment]
@FrankyShy 
BUT I can't install the pod since they are your private pod. Once I change it into `pod 'WCDB', '1.0.7.5'`, then nothing failed.

Further more, you have no need to specify the `WCDBOptimizedSQLCipher` since it's linked automatically by WCDB. Maybe it's the reason why it failed.

**Comment by RingoD** [comment]
@FrankyShy  I don't understand what you mean. If `Pod 'WCDB', '1.0.7.5'` is OK, then why not use it?

**Comment by FrankyShy** [comment]
Because WCDB '1.0.7.5'（source code） will cost more compile time about 60~80 seconds than use framework at the same condition.

This is compile time of our project including WCDB '1.0.7.5' frameworks.It cost 291.2 seconds.
<img width="977" alt="9b25314c6568d226fc930b1e92b43f2d" src="https://user-images.githubusercontent.com/15244415/48600884-c0030100-e9a8-11e8-8bdc-f300a1e525de.png">

This is compile time of our project including WCDB '1.0.7.5' source code.It costs 357.2 scends.
<img width="937" alt="98f996c4a562cb3fd573fcca2fbbb540" src="https://user-images.githubusercontent.com/15244415/48600888-c6917880-e9a8-11e8-9398-fdd91aae28b8.png">

**Comment by FrankyShy** [comment]
The build time with using source code more than using framework . We have hundreds of products, scripts run time(archieve ipa) is very important for us.

**Comment by lsqswl** [comment]
> @lsqswl Then you can use the solution 2.

@RingoD We don't have  other SQLite frameworks except SQLCipher. Just WCDB and FMDB.

**Comment by RingoD** [comment]
@FrankyShy  There are only two differences between v1.0.7 and v1.0.7.5, which is "Compatible with Swift 4.2" and "Enable FTS5". None of these will make the build time of ObjC project different. 
1. You should test about how long does the building process of WCDB cost, BUT not your whole project.
2. Since they are third party frameworks for you in Pods, why should you have to build it every time?

**Comment by RingoD** [comment]
@lsqswl  I think his problem is solved except he has more questions about the build time.

**Comment by lsqswl** [comment]
I want to know why the frameworks with FMDB and WCDB are in conflict. And how can we resolve it. We hope we can use frameworks which has the advantage of convenient use. Such as it can drag to the project.

**Comment by RingoD** [comment]
@lsqswl  I think it's already answered.

> You get this error because you are linking multiple SQLite libraries. One comes from WCDB, and the others maybe come from system framework or other linked library.
There are several ways to solve it:
>
> Add use_framworks! into you podfile to use WCDB as a dynamic library.
Remove other SQLite frameworks except SQLCipher inside WCDB.

**Comment by lsqswl** [comment]
I try the demo provided by @FrankyShy , it can not resolve the question with adding "use_framworks!". The project just add "FMDB" and "WCDB". You mean the "FMDB" framework and "WCDB" frameword cannot co-exist together, is it?

**Comment by RingoD** [comment]
@lsqswl  You are out of conversation. I have never said that WCDB and FMDB can't exist together. It can, except SQLite framework can't be duplicated.

As for the demo, removing deprecated code and add `use_framework!` into the podfile will solve all issues, which I already told.
The new podfile looks like:
```
project 'WCDBDemo'
platform :ios, '8.0'

use_frameworks!

target 'WCDBDemo' do
    pod 'WCDB', '1.0.7.5'
    pod 'FMDB'
end
```

Then the console tells:
```
2018-11-19 15:40:44.729708+0800 WCDBDemo[71079:854701] !!!dbPath = /Users/sanhuazhang/Library/Developer/CoreSimulator/Devices/8F7F7D05-B58D-477B-BA72-375AFAB15787/data/Containers/Data/Application/8B203B2B-BA1E-4404-89B3-5BDD028C3298/Documents/student.db
2018-11-19 15:40:44.733253+0800 WCDBDemo[71079:854701] create table success
```
which contains no error.

**Comment by lsqswl** [comment]
@RingoD The method which your provided is not frameworks, the "WCDB" and "FMDB" are source codes. If use source codes, I will not ask the question. Sorry, I hope you can help us to solve the problem that framework can not coexist. If it is a bad question for you, I'm not going to ask. Sorry!

**Comment by RingoD** [comment]
@lsqswl  I think you are still out of the conversation. The method we are talking about is for pod, but not source code. In addition, the solution doesn't matter about whether using source code or not.

If you are still confused, push your own reproducible demo.

**Comment by FrankyShy** [comment]
@RingoD 
 (1)I didn’t mean the v1.0.7 and v1.0.7.5 makes different build time.I  mean at the same version of WCDB,  framework and source code makes different build time.Using framework will need more build time.

(2)And my test condition  is  at the same project ,the only diffreence is using framework  of WCDB and using source code of WCDB. It should be almost same build time ，but indeed it causes different build time.

(3)We have hundreds of products, so we need to (archieve ipa)build it many times .

(4)Using framework of WCDB will make build time faster about 80 seconds at our project,but it can’t be coexist with FMDB which makes problem “library routine called out of sequence ”

**Comment by RingoD** [comment]
@FrankyShy  I don't think these tests are correct.
The routine of building project with WCDB source code is 1. compile WCDB source code and link it into framework. 2. build your own code with WCDB framework.
The routine of building project with WCDB framework is 1. build your own code with WCDB framework.
Note that the routine with framework version is part of the routine with source code version. There is no reason it will cost more time.

I think you can check if it is correct by Xcode 10 Building Timing Summary feature. Or you can push two projects that can show these differences.

**Comment by FrankyShy** [comment]
The source code podspec:
[WCDBOptimizedSQLCipher(sourceCode).podspec.zip](https://github.com/Tencent/wcdb/files/2602527/WCDBOptimizedSQLCipher.sourceCode.podspec.zip)

[WCDB(sourceCode).podspec.zip](https://github.com/Tencent/wcdb/files/2602528/WCDB.sourceCode.podspec.zip)

The framework podspec：

[WCDBOptimizedSQLCipher(framework).podspec.zip](https://github.com/Tencent/wcdb/files/2602529/WCDBOptimizedSQLCipher.framework.podspec.zip)

[WCDB(framework).podspec.zip](https://github.com/Tencent/wcdb/files/2602533/WCDB.framework.podspec.zip)

FMDB podspce：

[FMDB.podspec.zip](https://github.com/Tencent/wcdb/files/2602538/FMDB.podspec.zip)

**Comment by FrankyShy** [comment]
Such as the "bugly" pod is framework and the "WCDB" pod is source. We deal the "WCDB"(all of our libraries are dealed framework) as framework pod. It can reduce the build time.

We have hundards of project(about 1000 ipa), build time is important to us.

**Comment by RingoD** [comment]
@FrankyShy  Podspec and other zip files you posted make no sense. I need a reproducible project that can show the differences between source code version and framework version.

**Comment by RingoD** [comment]
@FrankyShy  Yes, you make this question clear enough. But we have not seen any theoretical analysis, test data, reproducible steps or more same reports from other developers about this issue. 
So at this moment, we don't have enough information to figure out the problem.

**Comment by lsqswl** [comment]
![image](https://user-images.githubusercontent.com/16233756/48998752-74620b80-f18f-11e8-96ce-2ddb3627a348.png)
![image](https://user-images.githubusercontent.com/16233756/48998759-7926bf80-f18f-11e8-9f29-62706630a66d.png)
![image](https://user-images.githubusercontent.com/16233756/48998764-7deb7380-f18f-11e8-8aef-249dcad08998.png)
![image](https://user-images.githubusercontent.com/16233756/48998770-80e66400-f18f-11e8-99e6-aea8eb33eb4a.png)


https://github.com/Tencent/wcdb/issues/342

I have posted the difference between source code and framework. But unfortunately, I don't know how to make him understand. 

![unnamed](https://user-images.githubusercontent.com/16233756/48998971-2568a600-f190-11e8-87d5-510a68f70465.png)

**Comment by lsqswl** [comment]
We will follow up progress at all times. Thank you @FrankyShy .  We'll hang up the question also. I have a hard time to met someone with the same problem .

**Comment by RingoD** [comment]
@lsqswl  I think I'm waiting for you and get no reply at issue 342.

> @lsqswl Did you try to upgrade to v1.0.7.4 in your Cocoapods? If it still exists in v1.0.7.4, please provide a sample project for us.

By the way, your issue is quite different from @FrankyShy 's. You'd better submit a new issue.

**Comment by lsqswl** [comment]
The question is caused by using FMDB framework and WCDB framework. I guess the FMDB framewok include the db library also. But I can not know why.

**Comment by RingoD** [comment]
@lsqswl  We all know this! So I told you several times that you are out of conversation.
Listen, forget this issue, focus on your own problem and we are on the step to solve your issue. Do the following steps:
1. Upgrade the version of WCDB and try again. The latest we released is v1.0.7.5. If failed, do the step 2.
2.  Create a new project with the latest version WCDB and FMDB and try again. If failed, submit a new issue with this test project.

**Comment by lsqswl** [comment]
Thank you. I tried but failed. Every time you publish a new version and I will try it. The demo provided by @FrankyShy is right. I'm sure it is my quesion. You haven't answered him yet. That's the answer I need.  Maybe I'm here to add to the mess. Forget me and my issue, focus on this issue.

**Comment by RingoD** [comment]
@lsqswl  His first question(with demo) is solved by changing WCDB version to v1.0.7.5 and add `use_framework!`, which is also your original question in issue 342.
Now he is trying to figure out why changing WCDB version to v1.0.7.5 with dynamic framework slow down the building process.

In order not to complicate the problem, I will not reply you at this issue anymore and I will work on your own issue. Please forgive me for being rude.
